services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/Pipe.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.services.pipeline;
+
+import tech.pegasys.pantheon.metrics.Counter;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Forms the connection between two pipeline stages. A pipe is essentially a blocking queue with the
+ * added ability to signal when no further input is available because the pipe has been closed or
+ * the pipeline aborted.
+ *
+ * <p>In most cases a Pipe is used through one of two narrower interfaces it supports {@link
+ * InputPipe} and {@link OutputPipe}. These are designed to expose only the operations relevant to
+ * objects either reading from or publishing to the pipe respectively.
+ *
+ * @param <T> the type of item that flows through the pipe.
+ */
+public class Pipe<T> implements InputPipe<T>, OutputPipe<T> {
+  private static final Logger LOG = LogManager.getLogger();
+  private final BlockingQueue<T> queue;
+  private final int capacity;
+  private final Counter itemCounter;
+  private final AtomicBoolean closed = new AtomicBoolean();
+  private final AtomicBoolean aborted = new AtomicBoolean();
+
+  public Pipe(final int capacity, final Counter itemCounter) {
+    queue = new ArrayBlockingQueue<>(capacity);
+    this.capacity = capacity;
+    this.itemCounter = itemCounter;
+  }
+
+  @Override
+  public boolean isOpen() {
+    return !closed.get() && !aborted.get();
+  }
+
+  /**
+   * Get the number of items that can be queued inside this pipe.
+   *
+   * @return the pipe's capacity.
+   */
+  public int getCapacity() {
+    return capacity;
+  }
+
+  @Override
+  public boolean hasRemainingCapacity() {
```

Comment:
If the pipe is not open, should this return 0 ??

Reply:
Also seems reasonable.